docs/CUSTOM_RULES_101.md

```diff
@@ -0,0 +1,529 @@
+
+```
+java.lang.IllegalStateException: At least one issue expected
+    at com.google.common.base.Preconditions.checkState(Preconditions.java:145)
+    at org.sonar.java.checks.verifier.CheckVerifier.assertMultipleIssue(CheckVerifier.java:166)
+    at org.sonar.java.checks.verifier.CheckVerifier.checkIssues(CheckVerifier.java:161)
+    at org.sonar.java.checks.verifier.JavaCheckVerifier.scanFile(JavaCheckVerifier.java:237)
+    at org.sonar.java.checks.verifier.JavaCheckVerifier.scanFile(JavaCheckVerifier.java:220)
+    at org.sonar.java.checks.verifier.JavaCheckVerifier.scanFile(JavaCheckVerifier.java:216)
+    at org.sonar.java.checks.verifier.JavaCheckVerifier.verify(JavaCheckVerifier.java:99)
+    at org.sonar.template.java.checks.MyFirstCustomCheckTest.test(MyFirstCustomCheckTest.java:10)
+    ...
+```
+
+### First version: Using syntax trees and API basics
+
+Before we start with the implementation of the rule itself, you need a little background.
+
+Prior to running any rule, the SonarQube Java Analyzer parses a given Java code file and produces an equivalent data structure: the **Syntax Tree**. Each construction of the Java language can be represented with a specific kind of Syntax Tree, detailing each of its particularities. Each of these constructions is associated with a specific Kind as well as an interface explicitly describing all its particularities. For instance, the kind associated to the declaration of a method will be `org.sonar.plugins.java.api.tree.Tree.Kind.METHOD`, and its interface defined by  `org.sonar.plugins.java.api.tree.MethodTree`. All the kinds are listed in the the [`Kind` enum of the Java Plugin](https://github.com/SonarSource/sonar-java/blob/5.12.1.17771/java-frontend/src/main/java/org/sonar/plugins/java/api/tree/Tree.java#L47).
+
+When creating the rule class, we chose to implement the `IssuableSubscriptionVisitor` class from the API. This class, on top of providing a bunch of useful methods to raise issues, also **defines the strategy which will be used when analyzing a file**. As its name is telling us, it is based on a subscription mechanism, allowing to specify on what kind of tree the rule should react. The list of node types to cover is specified through the `nodesToVisit()` method. In the previous steps, we modified the implementation of the method to return an empty list, therefore not subscribing to any node of the syntax tree.
+
+Now its finally time to jump in to the implementation of our first rule! Go back to the `MyFirstCustomCheck` class, and modify the list of Kinds returned by the nodesToVisit() method. Since our rule targets method declarations, we only need to visit methods. To do so, simply add `Kind.METHOD` as a parameter of the returned immutable list, as shown in the following code snippet.
+
+```java
+@Override
+public List<Kind> nodesToVisit() {
+  return ImmutableList.of(Kind.METHOD);
```

Comment:
I think we should avoid using Guava in the tutorial, let's use `Collections.singletonList` here